File: src/main/java/com/cj/messagestreaming/pubsub/PubSub.java
package com.cj.messagestreaming.pubsub;

import com.cj.messagestreaming.Types;
import com.spotify.google.cloud.pubsub.client.Message;
import com.spotify.google.cloud.pubsub.client.Publisher;
import com.spotify.google.cloud.pubsub.client.Pubsub;
import com.spotify.google.cloud.pubsub.client.Puller;

import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.stream.Stream;

import static com.spotify.google.cloud.pubsub.client.Puller.builder;

public class PubSub {

    public static Types.Subscription subscribe(PubSubConfig config) throws ExecutionException, InterruptedException {
        return subscribe(config, Pubsub.builder().build());
    }

    public static Types.Publication publish(PubSubConfig config) throws ExecutionException, InterruptedException {
        return publish(config, Pubsub.builder().build());
    }

    protected static Types.Subscription subscribe(PubSubConfig config, Pubsub pubsub) throws ExecutionException, InterruptedException {

        final String subscription = config.getSubscription();
        final String project = config.getProject();
        final String topic = config.getTopic();

        final List<byte[]> records = new ArrayList<>();

        pubsub.createSubscription(project, subscription, topic).get();

        final Puller.MessageHandler handler = (puller, sub, message, ackId) -> {
            records.add(message.decodedData());
            return CompletableFuture.completedFuture(ackId);
        };

        final Puller puller = builder()
                .pubsub(pubsub)
                .project(project)
                .subscription(subscription)
                .concurrency(32)
                .messageHandler(handler)
                .build();

        Stream<byte[]> results = records.stream();

        return null;
    }

    protected static Types.Publication publish(PubSubConfig config, Pubsub pubsub) throws ExecutionException, InterruptedException {

        final String topic = config.getTopic();
        final String project = config.getProject();

        pubsub.createTopic(project, topic).get();

        final Publisher publisher = Publisher.builder()
                .pubsub(pubsub)
                .project(project)
                .concurrency(128)
                .build();

        return (data) -> publisher.publish(topic, Message.of(Message.encode(data)));
    }
}

File: src/main/scala/com/cj/messagestreaming/pubsub/PubSub.scala
//package com.cj.messagestreaming.pubsub
//
//import java.util.concurrent.{CompletableFuture, CompletionStage}
//
//import com.cj.messagestreaming.{Publication, Subscription}
//import com.cj.collections.{IterableBlockingQueue, IteratorStream}
//import com.spotify.google.cloud.pubsub.client.Puller.MessageHandler
//import com.spotify.google.cloud.pubsub.client.{Message, Publisher, Pubsub, Puller}
//
//object PubSub {
//  case class PubSubConfig(project: String, topic: String)
//
//  def makeSubscription(config: PubSubConfig, subscriberId: String): Subscription = {
//    val pubsubClient = Pubsub.builder().build()
//
//    pubsubClient.createSubscription(config.project, config.topic)
//
//    val (enqueueMessage, stream) = subscribe()
//
//    val handler : Puller.MessageHandler = new MessageHandler {
//      override def handleMessage(puller: Puller, s: String, message: Message, ackId: String): CompletionStage[String] =
//        enqueueMessage(message,ackId)
//    }
//
//    Puller.builder()
//      .pubsub(pubsubClient)
//      .project(config.project)
//      .subscription(subscriberId)
//      .messageHandler(handler)
//      .build()
//
//    stream
//  }
//
//  type AckId = String
//  type PubSubCallback = (Message, AckId) => CompletionStage[String]
//
//  protected[pubsub] def subscribe() : (PubSubCallback, Subscription) = {
//    //the pubsubcallback writes to a queue
//    //and the subscription reads from the same queue
//    val q = new IterableBlockingQueue[Array[Byte]]()
//    val stream = new IteratorStream(q.iterator)
//
//    def callback = (message : Message, ackId : AckId) => {
//      q.add(message.decodedData())
//      CompletableFuture.completedFuture(ackId)
//    }
//
//    (callback, stream)
//  }
//
//  def makePublication(config: PubSubConfig): Publication = {
//    val pubsubClient = Pubsub.builder().build()
//
//    pubsubClient.createTopic(config.project, config.topic)
//
//    val publisher = Publisher.builder()
//      .pubsub(pubsubClient)
//      .project(config.project)
//      .build()
//
//    publish(config.topic, publisher)
//  }
//
//  protected[pubsub] def publish(topic: String, publisher: Publisher) : Publication = {
//    (data: Array[Byte]) => publisher.publish(topic, Message.of(Message.encode(data)))
//  }
//}
